GracefulShutdown upgrade may get stuck in between versions

There's a slight chance for 'TestOperandUpgrades' to fail in between individual version upgrades due to new StatefulSet version not being created.

**Operator log:**
[infinispan-operator-controller-manager-5776c64f76-l889k.log](https://github.com/user-attachments/files/19253863/infinispan-operator-controller-manager-5776c64f76-l889k.log)

**Test log:**
```
2025/03/14 19:22:43 INFO GET: https://test-operand-upgrades-external-datagrid.apps.paris.rhdg-qe.rhocf-dev.net:443/rest/v2/caches/volatileCache?action=size
2025/03/14 19:23:06 INFO GET: https://test-operand-upgrades-external-datagrid.apps.paris.rhdg-qe.rhocf-dev.net:443/rest/v2/container/health/status
2025/03/14 19:23:07 INFO GET: https://test-operand-upgrades-external-datagrid.apps.paris.rhdg-qe.rhocf-dev.net:443/rest/v2/caches/persistentCache?action=size
2025/03/14 19:23:07 INFO Next Operand 15.0.3
2025/03/14 19:23:09 INFO Upgrading Operand to 15.0.3
2025/03/14 19:23:38 INFO Upgrade complete
2025/03/14 19:23:42 INFO GET: https://test-operand-upgrades-external-datagrid.apps.paris.rhdg-qe.rhocf-dev.net:443/rest/v2/container/health/status
2025/03/14 19:23:42 INFO GET: https://test-operand-upgrades-external-datagrid.apps.paris.rhdg-qe.rhocf-dev.net:443/rest/v2/caches/persistentCache?action=size
2025/03/14 19:23:55 INFO GET: https://test-operand-upgrades-external-datagrid.apps.paris.rhdg-qe.rhocf-dev.net:443/rest/v2/caches/volatileCache?action=size
2025/03/14 19:24:14 INFO GET: https://test-operand-upgrades-external-datagrid.apps.paris.rhdg-qe.rhocf-dev.net:443/rest/v2/container/health/status
2025/03/14 19:24:14 INFO GET: https://test-operand-upgrades-external-datagrid.apps.paris.rhdg-qe.rhocf-dev.net:443/rest/v2/caches/persistentCache?action=size
2025/03/14 19:24:14 INFO Next Operand 15.0.4
2025/03/14 19:24:16 INFO Upgrading Operand to 15.0.4
2025/03/14 19:34:17 ERROR timed out waiting for the condition
2025/03/14 19:34:18 ERROR Encountered error: previous terminated container "infinispan" in pod "14-0-13" not found
2025/03/14 19:34:19 ERROR Encountered error: previous terminated container "infinispan" in pod "14-0-17" not found
...
2025/03/14 19:34:22 ERROR Encountered error: unable to write metrics to file for pod '15-0-3:'datagrid': pod not running. status.phase='Succeeded'
2025/03/14 19:34:22 ERROR Encountered error: unable to write metrics to file for pod 'upgrade-backup:'datagrid': pod not running. status.phase='Succeeded'
goroutine 303 [running]:
runtime/debug.Stack()
	/usr/lib/golang/src/runtime/debug/stack.go:24 +0x5e
github.com/infinispan/infinispan-operator/test/e2e/utils.TestKubernetes.CleanNamespaceAndLogWithPanic({0xc000d8ecc0}, 0xc0000d2820, {0xc00004c012, 0x8}, {0x15ca3a0, 0xc00058eee0})
	/home/pdrobek/git/infinispan-operator/test/e2e/utils/kubernetes.go:226 +0x1825
github.com/infinispan/infinispan-operator/test/e2e/utils.TestKubernetes.CleanupOLMTest({0x43fce0?}, 0xc0000d2820, {0xc000c22618, 0x13}, {0x186a191, 0x13}, {0xc00004a197, 0x13}, {0xc00004c035, 0xa})
	/home/pdrobek/git/infinispan-operator/test/e2e/utils/olm.go:178 +0x178
panic({0x15ca3a0?, 0xc00058eee0?})
	/usr/lib/golang/src/runtime/panic.go:770 +0x132
github.com/infinispan/infinispan-operator/test/e2e/utils.ExpectNoError(...)
	/home/pdrobek/git/infinispan-operator/test/e2e/utils/asserts.go:21
github.com/infinispan/infinispan-operator/test/e2e/utils.TestKubernetes.WaitForInfinispanStateWithTimeout({0xc000123520?}, {0xc000b9bb78, 0x15}, {0xc000bfbb48, 0x8}, 0x8bb2c97000, 0xc0006c3e30)
	/home/pdrobek/git/infinispan-operator/test/e2e/utils/kubernetes.go:697 +0x116
github.com/infinispan/infinispan-operator/test/e2e/utils.TestKubernetes.WaitForInfinispanState(...)
	/home/pdrobek/git/infinispan-operator/test/e2e/utils/kubernetes.go:682
github.com/infinispan/infinispan-operator/test/e2e/upgrade.TestOperandUpgrades(0xc0000d2820)
	/home/pdrobek/git/infinispan-operator/test/e2e/upgrade/upgrade_operands_test.go:88 +0xcb1
testing.tRunner(0xc0000d2820, 0x194a750)
	/usr/lib/golang/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 1
	/usr/lib/golang/src/testing/testing.go:1742 +0x390

    kubernetes.go:227: timed out waiting for the condition 
        
2025/03/14 19:34:29 ERROR Encountered error: previous terminated container "manager" in pod "infinispan-operator-controller-manager-5776c64f76-l889k" not found
2025/03/14 19:34:30 ERROR Encountered error: previous terminated container "olm-operator" in pod "olm-operator-5569b74676-mdggw" not found
2025/03/14 19:34:30 ERROR Encountered error: previous terminated container "catalog-operator" in pod "catalog-operator-bbcfb64d6-lnsdn" not found
--- FAIL: TestOperandUpgrades (1691.07s)
```

--- a/api/v2alpha1/batch_webhook_test.go
+++ b/api/v2alpha1/batch_webhook_test.go
@@ -63,7 +63,7 @@ var _ = Describe("Batch Webhook", func() {
 				},
 				Spec: BatchSpec{
 					Cluster: "some-cluster",
-					Config:  pointer.String("create cache --template=org.infinispan.DIST_SYNC batch-cache"),
+					Config:  pointer.String("create counter --concurrency-level=1 --initial-value=5 --storage=VOLATILE --type=weak batch-counter"),
 				},
 			}
 
@@ -112,7 +112,7 @@ var _ = Describe("Batch Webhook", func() {
 				},
 				Spec: BatchSpec{
 					Cluster: "some-cluster",
-					Config:  pointer.String("create cache --template=org.infinispan.DIST_SYNC batch-cache"),
+					Config:  pointer.String("create counter --concurrency-level=1 --initial-value=5 --storage=VOLATILE --type=weak batch-counter"),
 				},
 			}
 
@@ -144,7 +144,7 @@ var _ = Describe("Batch Webhook", func() {
 				},
 				Spec: BatchSpec{
 					Cluster: "some-cluster",
-					Config:  pointer.String("create cache --template=org.infinispan.DIST_SYNC batch-cache"),
+					Config:  pointer.String("create counter --concurrency-level=1 --initial-value=5 --storage=VOLATILE --type=weak batch-counter"),
 					Container: &BatchContainerSpec{
 						Memory: "1Gi:5Gi",
 						CPU:    "1000m:2000m",

--- a/config/samples/batch/infinispan_v2alpha1_batch.yaml
+++ b/config/samples/batch/infinispan_v2alpha1_batch.yaml
@@ -5,6 +5,4 @@ metadata:
 spec:
   cluster: example-infinispan
   config: |
-    create cache --template=org.infinispan.DIST_SYNC mycache
-    put --cache=mycache hello world
-    put --cache=mycache hola mundo
+    create counter --concurrency-level=1 --initial-value=5 --storage=VOLATILE --type=weak batch-counter

--- a/config/samples/cache/infinispan_v2alpha1_cache.yaml
+++ b/config/samples/cache/infinispan_v2alpha1_cache.yaml
@@ -5,4 +5,4 @@ metadata:
 spec:
   clusterName: example-infinispan
   name: mycache
-  templateName: org.infinispan.DIST_SYNC
+  template: <distributed-cache />

--- a/documentation/asciidoc/topics/proc_creating_caches_hotrod.adoc
+++ b/documentation/asciidoc/topics/proc_creating_caches_hotrod.adoc
@@ -32,7 +32,6 @@ Add cache configuration to `hotrod-client.properties` as in the following exampl
 [source,options="nowrap",subs=attributes+]
 ----
 # Add cache configuration
-infinispan.client.hotrod.cache.my-cache.template_name=org.infinispan.DIST_SYNC
-infinispan.client.hotrod.cache.another-cache.configuration=<infinispan><cache-container><distributed-cache name=\"another-cache\"/></cache-container></infinispan>
-infinispan.client.hotrod.cache.my-other-cache.configuration_uri=file:/path/to/configuration.xml
+infinispan.client.hotrod.cache.my-cache.configuration=<infinispan><cache-container><distributed-cache name=\"my-cache\"/></cache-container></infinispan>
+infinispan.client.hotrod.cache.another-cache.configuration_uri=file:/path/to/configuration.xml
 ----

--- a/documentation/asciidoc/topics/yaml/batch_inline.yaml
+++ b/documentation/asciidoc/topics/yaml/batch_inline.yaml
@@ -5,6 +5,5 @@ metadata:
 spec:
   cluster: {example_crd_name}
   config: |
-    create cache --template=org.infinispan.DIST_SYNC mycache
-    put --cache=mycache hello world
-    put --cache=mycache hola mundo
+    create counter --concurrency-level=1 --initial-value=5 --storage=VOLATILE --type=weak batch-counter
+

--- a/documentation/asciidoc/topics/yaml/cache_template.yaml
+++ b/documentation/asciidoc/topics/yaml/cache_template.yaml
@@ -1,8 +1,0 @@
-apiVersion: infinispan.org/v2alpha1
-kind: Cache
-metadata:
-  name: mycachedefinition
-spec:
-  clusterName: {example_crd_name}
-  name: mycache
-  templateName: org.infinispan.DIST_SYNC

--- a/test/e2e/batch/batch_helper.go
+++ b/test/e2e/batch/batch_helper.go
@@ -2,6 +2,9 @@ package batch
 
 import (
 	"fmt"
+	ispnv1 "github.com/infinispan/infinispan-operator/api/v1"
+	batchCtrl "github.com/infinispan/infinispan-operator/controllers"
+	corev1 "k8s.io/api/core/v1"
 	"testing"
 	"time"
 
@@ -58,4 +61,23 @@ func (b BatchHelper) WaitForValidBatchPhase(name string, phase v2.BatchPhase) *v
 	}
 	tutils.ExpectNoError(err)
 	return batch
+}
+
+func (b BatchHelper) CreateBatchCM(infinispan *ispnv1.Infinispan) *corev1.ConfigMap {
+	configMapName := infinispan.Name + "-cm"
+	data := map[string]string{
+		batchCtrl.BatchFilename: "create cache --file=/etc/batch/mycache.xml mycache",
+		"mycache.xml":           "<distributed-cache />",
+	}
+
+	configMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configMapName,
+			Namespace: infinispan.Namespace,
+		},
+		Data: data,
+	}
+
+	b.testKube.CreateConfigMap(configMap)
+	return configMap
 }

--- a/test/e2e/batch/batch_test.go
+++ b/test/e2e/batch/batch_test.go
@@ -16,8 +16,6 @@ import (
 	"github.com/infinispan/infinispan-operator/pkg/infinispan/client/api"
 	tutils "github.com/infinispan/infinispan-operator/test/e2e/utils"
 	batchv1 "k8s.io/api/batch/v1"
-	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/wait"
 )
 
@@ -48,7 +46,6 @@ func testBatchInlineConfig(t *testing.T, infinispan *v1.Infinispan) {
 
 	httpClient := tutils.HTTPClientForCluster(infinispan, testKube)
 	ispn := ispnClient.New(tutils.CurrentOperand, httpClient)
-	assertCacheExists("batch-cache", ispn)
 	assertCounterExists("batch-counter", ispn)
 	testKube.DeleteBatch(batch)
 	waitForK8sResourceCleanup(name)
@@ -59,22 +56,10 @@ func TestBatchConfigMap(t *testing.T) {
 	defer testKube.CleanNamespaceAndLogOnPanic(t, tutils.Namespace)
 
 	infinispan := createCluster(t)
-
-	configMapName := infinispan.Name + "-cm"
-	configMap := &corev1.ConfigMap{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      configMapName,
-			Namespace: infinispan.Namespace,
-		},
-		Data: map[string]string{
-			batchCtrl.BatchFilename: batchString(),
-		},
-	}
-
-	testKube.CreateConfigMap(configMap)
+	configMap := helper.CreateBatchCM(infinispan)
 	defer testKube.DeleteConfigMap(configMap)
 
-	batch := helper.CreateBatch(t, infinispan.Name, infinispan.Name, nil, &configMapName, nil)
+	batch := helper.CreateBatch(t, infinispan.Name, infinispan.Name, nil, &(configMap.Name), nil)
 
 	helper.WaitForValidBatchPhase(infinispan.Name, v2.BatchSucceeded)
 	testKube.DeleteBatch(batch)
@@ -82,8 +67,7 @@ func TestBatchConfigMap(t *testing.T) {
 
 	httpClient := tutils.HTTPClientForCluster(infinispan, testKube)
 	ispn := ispnClient.New(tutils.CurrentOperand, httpClient)
-	assertCacheExists("batch-cache", ispn)
-	assertCounterExists("batch-counter", ispn)
+	assertCacheExists("mycache", ispn)
 }
 
 func TestBatchFail(t *testing.T) {
@@ -126,8 +110,7 @@ func TestBatchWithResources(t *testing.T) {
 }
 
 func batchString() string {
-	batchScript := `create cache --template=org.infinispan.DIST_SYNC batch-cache
-	create counter --concurrency-level=1 --initial-value=5 --storage=VOLATILE --type=weak batch-counter`
+	batchScript := `create counter --concurrency-level=1 --initial-value=5 --storage=VOLATILE --type=weak batch-counter`
 	return strings.ReplaceAll(batchScript, "\t", "")
 }
 

--- a/test/e2e/cache/cache_test.go
+++ b/test/e2e/cache/cache_test.go
@@ -81,13 +81,8 @@ func TestCacheCR(t *testing.T) {
 		assertConfigListenerHasNoErrorsOrRestarts(t, ispn)
 	}
 
-	//Test for CacheCR with TemplateName
-	cache := cacheCR("cache-with-static-template", ispn)
-	cache.Spec.TemplateName = "org.infinispan.DIST_SYNC"
-	test(cache)
-
 	//Test for CacheCR with TemplateXML
-	cache = cacheCR("cache-with-xml-template", ispn)
+	cache := cacheCR("cache-with-xml-template", ispn)
 	cache.Spec.Template = "<distributed-cache name=\"cache-with-xml-template\" mode=\"SYNC\"/>"
 	test(cache)
 

--- a/test/e2e/upgrade/common.go
+++ b/test/e2e/upgrade/common.go
@@ -248,12 +248,14 @@ func checkServicePorts(t *testing.T, name string) {
 	assert.Equal(t, constants.InfinispanAdminPort, int(adminPort[0].Port))
 }
 
-func checkBatch(t *testing.T, name string) {
+func checkBatch(t *testing.T, infinispan *ispnv1.Infinispan) {
 	// Run a batch in the migrated cluster
 	batchHelper := batchtest.NewBatchHelper(testKube)
-	config := "create cache --template=org.infinispan.DIST_SYNC batch-cache"
-	batch := batchHelper.CreateBatch(t, name, name, &config, nil, nil)
-	batchHelper.WaitForValidBatchPhase(name, v2.BatchSucceeded)
+	configMap := batchHelper.CreateBatchCM(infinispan)
+	defer testKube.DeleteConfigMap(configMap)
+
+	batch := batchHelper.CreateBatch(t, infinispan.Name, infinispan.Name, nil, &(configMap.Name), nil)
+	batchHelper.WaitForValidBatchPhase(infinispan.Name, v2.BatchSucceeded)
 	testKube.DeleteBatch(batch)
 }
 

--- a/test/e2e/upgrade/upgrade_operands_test.go
+++ b/test/e2e/upgrade/upgrade_operands_test.go
@@ -109,7 +109,7 @@ func TestOperandUpgrades(t *testing.T) {
 
 		tutils.NewCacheHelper(volatileCacheName, client).AssertSize(numEntries)
 		checkServicePorts(t, ispn.Name)
-		checkBatch(t, ispn.Name)
+		checkBatch(t, ispn)
 
 		// Kill the first pod to ensure that the cluster can recover from failover after upgrade
 		err := testKube.Kubernetes.Client.Delete(ctx, &corev1.Pod{

--- a/test/e2e/upgrade/upgrade_test.go
+++ b/test/e2e/upgrade/upgrade_test.go
@@ -160,7 +160,7 @@ func TestUpgrade(t *testing.T) {
 	}
 
 	checkServicePorts(t, spec.Name)
-	checkBatch(t, spec.Name)
+	checkBatch(t, spec)
 
 	// Kill the first pod to ensure that the cluster can recover from failover after upgrade
 	err := testKube.Kubernetes.Client.Delete(ctx, &corev1.Pod{

